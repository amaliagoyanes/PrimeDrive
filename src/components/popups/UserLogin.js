import axios from 'axios';
import React, {Component} from 'react';
import {
    Dimensions, AsyncStorage,
    TouchableOpacity,
} from 'react-native';
import {Row} from 'react-native-easy-grid';
import Modal from 'react-native-root-modal';
import {scale} from '../../scaling';
import {fetchData} from '../../api';
import firestack, {urlBase} from '../../fbconfig.js'

import {UserLoginBlock, UserSearchBlock, ModalHeader, ModalCloseButton, } from '../index';
import {updateUserUID, updatePermissions} from '../../reducers/actions';

import styles from '../../style/styles';


export default class UserLogin extends Component {
    constructor(props){
        super(props);
        this.state = {
            open: false, data: [], processing: false
        };

        this.modalName = 'app.welcomeUpper';

        this.close = this.close.bind(this);
        this.open = this.open.bind(this);
        this.login = this.login.bind(this);
        this.getData = this.getData.bind(this);
    }

    getData() {
        fetchData('users', {role: 'user'}).then(result => {
            this.setState({data: result, open: true});
        });
    }

    open(){
        this.setState({open: true}, () => {
            if (this.state.data.length === 0) {
                this.getData();
            }
        });
    }

    close(){
        this.setState({open: false})
    }

    login(number) {
        this.setState({processing: true});
        firestack.auth.getToken().then(res => {
            let url = `${urlBase}/develop/auth-by-number`;
            let token = res.token;
            let instance = axios.create({
                headers: {idtoken: token},
            });
            console.log('NUMBER', number);
            let data = {number: number};
            instance.post(url, data)
                .then(response => {
                    this.setState({processing: false});
                    this.close();
                    AsyncStorage.setItem('@PrimeDrive:user_info',
                        JSON.stringify({number: response.data.db_data.number, name: response.data.db_data.name,
                            permissions: response.data.db_data.permissions}))
                        .then(() => {
                            this.props.dispatch(updateUserUID(response.data.auth_data.uid));
                            this.props.dispatch(updatePermissions(response.data.db_data.permissions || {}));
                        });
                    AsyncStorage.setItem('@PrimeDrive:user_uid', response.data.auth_data.uid);
                    AsyncStorage.getItem('@PrimeDrive:last_shift_uid').then((value) => {
                        if (!value) {
                            this.props.openShift();
                        }
                    });
                }).catch(err => {
                console.log('lambda post', err);
                this.setState({processing: false, error: true});
            })
        }).catch(err => {
            console.log('auth token', err);
            this.setState({processing: false, error: true});
        });
    }


    render() {
        return (
            <Modal
                onRequestClose={this.close}
                visible={this.state.open}
                style={styles.modal}
            >
                <TouchableOpacity style={styles.overlay} onPress={this.close}>
                    <TouchableOpacity activeOpacity={1} onPress={(event) => event.stopPropagation()} style={{
                        position: 'absolute',
                        top: Dimensions.get('screen').height * 0.1,
                        left: Dimensions.get('screen').width * 0.1,
                        width: Dimensions.get('screen').width * 0.8,
                        height: Dimensions.get('screen').height * 0.8,
                        backgroundColor: '#e8eaeb',
                        paddingLeft: scale(40),
                        paddingRight: scale(40),
                        paddingTop: scale(0),
                        paddingBottom: scale(20)
                    }}>
                        <ModalCloseButton close={this.close}/>
                        <ModalHeader modalName={this.modalName} intl={this.props.intl}/>
                        <Row size={20}>
                            <UserSearchBlock data={this.state.data}
                                             login={(id) => this.login(id)}/>

                            <UserLoginBlock token={this.props.token} close={this.close} login={this.login}
                                            ref={elt => this._login = elt} intl={this.props.intl} processing={this.state.processing}
                                            openShift={this.props.openShift} dispatch={this.props.dispatch} />
                        </Row>
                        <Row size={0.5}/>
                    </TouchableOpacity>
                </TouchableOpacity>
            </Modal>
        )
    }
}
