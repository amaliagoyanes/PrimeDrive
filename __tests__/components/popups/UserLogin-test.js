'use strict';

import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import UserLogin from '../../../src/components/popups/UserLogin';


it('renders correctly', () => {
    const tree = shallow(
        <UserLogin />);
    tree.setState({open: true});
    expect(tree).toMatchSnapshot();
    expect(tree.find('UserSearchBlock').length).toEqual(1);
    expect(tree.find('UserLoginBlock').length).toEqual(1);
});

it('has modal header and close button', () => {
    const tree = shallow(
        <UserLogin />);
    tree.setState({open: true});
    expect(tree.find('ModalCloseButton').length).toEqual(1);
    expect(tree.find('ModalHeader').length).toEqual(1);
});