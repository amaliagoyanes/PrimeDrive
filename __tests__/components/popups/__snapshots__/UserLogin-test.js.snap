// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Modal
  onRequestClose={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(255, 255, 255, 0.2)",
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    }
  }
  visible={true}
>
  <TouchableOpacity
    activeOpacity={0.2}
    focusedOpacity={0.7}
    onPress={[Function]}
    style={
      Object {
        "backgroundColor": "rgba(255,255,255,0.1)",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={1}
      focusedOpacity={0.7}
      onPress={[Function]}
      style={
        Object {
          "backgroundColor": "#e8eaeb",
          "height": 1067.2,
          "left": 75,
          "paddingBottom": 20,
          "paddingLeft": 40,
          "paddingRight": 40,
          "paddingTop": 20,
          "position": "absolute",
          "top": 133.4,
          "width": 600,
        }
      }
    >
      <ModalCloseButton
        close={[Function]}
      />
      <ModalHeader
        modalName="WELCOME"
      />
      <RowNB
        size={20}
      >
        <UserSearchBlock
          login={[Function]}
        />
        <UserLoginBlock
          close={[Function]}
        />
      </RowNB>
      <RowNB
        size={0.5}
      />
    </TouchableOpacity>
  </TouchableOpacity>
</Modal>
`;
